CODING_STYLE.md

```diff
@@ -0,0 +1,199 @@
+2. Leading "c_" to const variables (unless part of an external API).
+3. Leading "g_" to global (non-const) variables.
+4. Leading "s_" to static (non-const, non-global) variables.
+
+
+
+## Error reporting
+
+Prefer exception to bool/int return type.
+
+
+## Declarations
+
+1. {Typename} + {qualifiers} + {name}. (**TODO**: Against [NL.26](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#nl26-use-conventional-const-notation))
+2. Only one per line.
+3. Favour declarations close to use; don't habitually declare at top of scope ala C.
+4. Always pass non-trivial parameters with a const& suffix.
+5. If a function returns multiple values, use std::tuple (std::pair acceptable).
+   Prefer not using */& arguments, except where efficiency requires.
+   (**TODO**: Compare with [C++ Core Guidelines F.21](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#f21-to-return-multiple-out-values-prefer-returning-a-tuple-or-struct) )
+6. Never use a macro where adequate non-preprocessor C++ can be written.
+7. Make use of auto whenever type is clear or unimportant:
+   - Always avoid doubly-stating the type.
+   - Use to avoid vast and unimportant type declarations.
+   - However, avoid using auto where type is not immediately obvious from the context, and especially not for arithmetic expressions.
+8. Don't pass bools: prefer enumerations instead.
+9. Prefer enum class to straight enum.
+
+
+    // WRONG:
+    const double d = 0;
+    int i, j;
+    char *s;
+    float meanAndSigma(std::vector<float> _v, float* _sigma, bool _approximate);
+    Derived* x(dynamic_cast<Derived*>(base));
+    for (map<ComplexTypeOne, ComplexTypeTwo>::iterator i = l.begin(); i != l.end(); ++l) {}
+
+
+    // CORRECT:
+    enum class Accuracy
+    {
+        Approximate,
+        Exact
+    };
+    double const d = 0;
+    int i;
+    int j;
+    char* s;
+    std::tuple<float, float> meanAndSigma(std::vector<float> const& _v, Accuracy _a);
+    auto x = dynamic_cast<Derived*>(base);
+    for (auto i = x.begin(); i != x.end(); ++i) {}
+
+
+## Structs & classes
+
+1. Structs to be used when all members public and no virtual functions.
+   - In this case, members should be named naturally and not prefixed with 'm_'
+2. Classes to be used in all other circumstances.
+
+
+
+## Members
+
+1. One member per line only.
+2. Private, non-static, non-const fields prefixed with m_.
+3. Avoid public fields, except in structs.
+4. Use override, final and const as much as possible.
+5. No implementations with the class declaration, except:
+   - template or force-inline method (though prefer implementation at bottom of header file).
+   - one-line implementation (in which case include it in same line as declaration).
+6. For a property 'foo'
+   - Member: m_foo;
+   - Getter: foo(); also: for booleans, isFoo()
+   - Setter: setFoo();
+
+
+## Naming
+
+1. Collection conventions:
+   - -s means std::vector e.g. using MyTypes = std::vector<MyType>
```

Comment:
put these things in backticks, otherwise names inside angle brackets are not rendered